src/runner.rs

```diff
@@ -6,511 +6,180 @@
-        use example_result;
-
-        let res = catch_unwind(AssertUnwindSafe(|| test_function()));
-
-        res.unwrap_or_else(|_| example_result::FAILED_RES)
-    }
-
-    fn run_and_recurse(report: &mut TestReport,
-                       child_ctx: &mut Context,
-                       handlers: &mut Handlers)
-                       -> ExampleResult {
-
-        use std::mem;
-        let mut result = example_result::SUCCESS_RES;
-
-        // As Runner::run consomes the runner, we can deconstruct the context tree as we travel
-        // among it.
-        // This effectively takes ownership of the context node and replace it with an empty one.
-        let ctx = mem::replace(child_ctx, Context::default());
-
-        let tests = ctx.tests;
-        let mut before_functions = ctx.before_each;
-        let mut after_functions = ctx.after_each;
-
-
-        for mut test_function in tests {
-            let test_res = {
-                for before_function in before_functions.iter_mut() {
-                    before_function()
-                }
-                let res = match test_function {
-                    Testable::Test(name, test_function) => {
-                        handlers.broadcast(&Event::StartTest(name));
-                        let res = Runner::run_test(test_function);
-                        handlers.broadcast(&Event::EndTest(res));
-                        res
-                    }
-                    Testable::Describe(ref name, ref mut desc) => {
-                        handlers.broadcast(&Event::StartDescribe(name.clone()));
-                        let res = Runner::run_and_recurse(report, desc, handlers);
-                        handlers.broadcast(&Event::EndDescribe);
-                        res
-                    }
-                };
-                for after_function in after_functions.iter_mut() {
-                    after_function()
-                }
-                res
-            };
-
-            report.total_tests += 1;
-
-            if test_res.is_ok() {
-                report.success_count += 1;
-            } else {
-                report.error_count += 1
-            }
-
-            result = test_res.or(result);
+impl Runner {
+    pub fn new(configuration: Configuration, handlers: Vec<Arc<Mutex<EventHandler>>>) -> Runner {
+        Runner {
+            configuration: configuration,
+            handlers: handlers,
         }
-
-        result
     }
+}
 
-    pub fn run(mut self) -> RunnerResult {
-        self.handlers.broadcast(&Event::StartRunner);
-
-        let mut report = TestReport::default();
-        let result = Runner::run_and_recurse(&mut report, &mut self.describe, &mut self.handlers);
-        let result = result.res().and(Ok(report)).or_else(|_| Err(report));
-
-        self.handlers.broadcast(&Event::FinishedRunner(result));
-        result
+impl Runner {
+    pub fn run<T>(self, suite: (Suite<T>, T)) -> SuiteReport
```

Comment:
One takes an explicit configuration, one uses the implicit default.

However given that test initiation has become quite cumbersome with the latest changes it doesn't make much a change, I guess. Maybe just drop the default config one and remove the `_with` from the explicitly configured variant.

My plan is to at some point (i.e. once the overall architecture has settled) re-introduce some API akin to `rdescribe(…)`, which would wrap all this:

```rust
let simple = rspec::formatter::Simple::new(io::stdout());
let formatter = Arc::new(Mutex::new(simple));
let configuration = Configuration::default(); // .parallel(false);
let runner = Runner::new(configuration, vec![formatter]);

#[derive(Clone, Debug)]
struct Environment {
    // …
}

let environment = Environment {
    // …
};

runner.run_or_exit(rspec::given("a BTreeSet", environment, |ctx| {
    // …
}));
```

into something more palatable, like this:

```rust
#[derive(Clone, Debug, Rspec)]
struct Environment {
    // …
}

rspec::run_or_exit(rspec::given("a BTreeSet", Environment {
    // …
}, |ctx| {
    // …
}));
```